RFD 0162: Machine ID Fleet Management, `BotInstance` resource and  other `token` joining improvements (#36510)
gravitational/teleport: rfd/0162-machine-id-token-join-method-bot-instance.md

- When using the `token` join method, a token can only be used once. This means
  creating hundreds of join tokens and managing securely distributing these to
  hosts.
- When managing a large fleet of `tbot` deployments, there is no way to
  track these within Teleport. This makes it more difficult to identify hosts
  which may need updating.

As we look to onboard more Enterprise customers to Machine ID, the pains of
the on-prem experience have become more apparent. Enterprise customers are
more likely to have on-prem deployments and these are likely to be larger in
scale.

## Details

### Current State

Currently, the `token` join method introduces a generation counter as a label
on the Bot user. This counter is contained within the Bot certificate and on
each renewal, this counter is incremented. When the counter within the certificate
de-synchronises with the counter on the user, the Bot is locked out as a security
measure.

The fact that this counter is stored within a label on the Bot user creates a 
one-to-one binding between a single instance of `tbot` and a single Bot user.
This is not the case when using the delegated join methods.
 

> ### Persistent Bot Instance Identity

Reviewer: After some experimentation, I think we'll actually need to do most of the more complex work involved in UUID verification even if we pick public key identifiers.

The problem, I think, is that `RegisterUsingTokenRequest` doesn't check the public key regardless, so clients could just lie about it. This means a malicious bot could potentially impersonate an existing instance. I don't think that's a hugely critical vulnerability but might help a malicious bot disguise itself.

I think we'll want to pass along an existing certificate to verify regardless... and we have to verify private key ownership, so that means we still need to return the new certs encrypted with the provided public key, after verifying it against auth's CA.

With that in mind, I don't think it's much of a stretch to include additional fields in the cert if we want.

It might honestly also be worthwhile to look at adding a grpc entrypoint to delegated joining. We'd then get a proper TLS session with the existing client cert which is IMO much less sketchy than manually verifying the certs and re-encrypting the payload from the existing endpoint.

Author: Yeah - that's a good point. Although, even if we pursue the simple model of encrypting the returned certs, for the purposes of the audit logs, a malicious actor can still pretend to be another bot instance. I thought about requiring a signature when making the initial request but I worry about how those can relatively easily be re-sent unless the signed value is some sort of challenge - perhaps use a timestamp ???

As you say, a gRPC entrypoint would resolve most of these concerns and would allow for any challenge and response flows we wanted. It'd also be an easy way to flip between the old and new behaviour in regards to BotInstances. I think the only challenge worth bearing in mind with a gRPC entrypoint is how an unauthenticated client would reach the auth server, today, the IAM and Azure join methods are terminated at the Proxy and the gRPC requests are re-made, but this would conceal the TLS certificate from the auth server. Perhaps that's not really a problem and unauthenticated requests could be made to the proxy directly, and when you have an existing cert, you can tunnel through and hit the auth server directly.

I like the idea generally and moving away from HTTP and inline with other efforts - my only major concern with moving this to gRPC would be avoiding scope creep. There's a lot in the RegisterUsingToken flow that people will want to refactor and change, and we'll need to resist that if we want to avoid this becoming a rather large task. Perhaps we set out a bounds and say for the V1 of the gRPC method, we'll mimic the HTTP fields precisely.

Reviewer: >  worry about how those can relatively easily be re-sent unless the signed value is some sort of challenge - perhaps use a timestamp ???

Yep, I tried to game out what a proper challenge would look like and was not liking the amount of effort or complexity it would add given it still feels like a hack vs a gRPC registration entrypoint.

> I like the idea generally and moving away from HTTP and inline with other efforts - my only major concern with moving this to gRPC would be avoiding scope creep

:+1: I'll pivot my investigation today to gauge how complex a gRPC refresh would be.

To cut scope initially I'd potentially be okay with a gRPC endpoint that is _only_ accessible by already authenticated users. Clients would onboard through HTTP, but renew via gRPC. I'll experiment and see what makes sense, if anything.

Author: > Yep, I tried to game out what a proper challenge would look like and was not liking the amount of effort or complexity it would add given it still feels like a hack vs a gRPC registration entrypoint.

Yeah the only thing that comes to mind is signing a timestamp and having the auth verify that it falls within a 5 minute window - but that's not fantastic for a bunch of reasons.

> 👍 I'll pivot my investigation today to gauge how complex a gRPC refresh would be.

Sounds good - gRPC gives us the option to do call and responses which is great. The TPM joining I did fairly recently is a nice-ish example of how to achieve this and does show how to use the existing JoinUsingTokenRequest proto within a challenge-response flow.

Reviewer: Alright, I'll be trying to merge our discussions back into the RFD today. My summary so far is:
* We'd ideally like to use UUIDs (or similar) as the primary bot identifier, to allow (among other things) key rotation; I'll rewrite this section to list that as the preferred method and public keys as the alternate
* I'll also add a short section on trust of this new identifier to help outline why I think it's important to verify it during renewal.

  individual Bot instances in use. We cannot easily determine if it's a single
  very active Bot instance, or many less active Bot instances.

To rectify this, a unique identifier should be established for an instance of
a Bot.

#### Public Key Fingerprint

One option is to modify the behaviour of `tbot` to persist and reuse the
keypair across renewals. We could then use a fingerprint of the public key as a
unique identifier of the Bot instance.

This feels like a natural identifier. It avoids introducing a new attribute to
certificates as the public key is already encoded within certificates. The
nature of public-key cryptography also means that this provides the Bot instance
a way to identify itself without needing an issued certificate.

To mitigate the risk of pre-image attacks, SHA256 will be used to determine the
fingerprint of the public key. In addition, the full public key should be
recorded and verified against when authenticating a Bot instance action.

It should be noted that with this technique, rotating the keypair of a `tbot`
instance would reset the identity of that instance. Rotation of this keypair

> would be unusual and this side effect seems expected.

Author: > We do create fresh keys for impersonated identities on each renewal today.

I lean towards this being slightly different - the impersonated identities go to an output which by it's very nature will be less secure than the bot's own internal storage. You may have a number of consumers that pull these secrets out into different places.

> would be unusual and this side effect seems expected.

Reviewer: Agreed, though while it's certainly _less_ pressing to rotate bots' internal keypairs, it's arguably a bit weird that we just never do it today. In other words, I think private key rotation is a reasonable thing to implement in the future, and some dedicated identifier would let us maintain a single `BotInstance` through that.

On the other hand, the negative consequence to rotating if we used public key identifiers would just be a new bot instance. It's convenient to have a clear lineage, but probably not critical.

##### Alternative: UUID Certificate Attribute

On the initial join of a Bot instance, we could generate a UUID to identify that 
Bot instance and encode this within the certificate. Upon renewals, the UUID
would be copied from the current certificate and into the new one.

> Whilst this is fairly easy to implement for the `token` join method, one

Reviewer 2: This limitation would go away if we moved the mechanism for delegated joins over to this mechanism, correct?

challenge for the delegated join methods is that rather than renewing, the
`tbot` instance merely re-joins. As the join RPCs are unauthenticated, the 
previous certificate of the Bot instance is not readily available. We can
either:

- Accept this limitation and treat each renewal of a delegated Bot instance as a
  new Bot instance. This is likely unacceptable and would limit any advantages

>   of this work to the `token` join method.

Author: > I'd expect all delegated joins from CI providers to be new bot instances every time, since they're stateless.

Hmm, I'm not sure I'd necessarily want to set that precedent. Even in CI situations, a long-running CI job may renew several times (e.g a background tbot database tunnel with a CI job running a long migration or an ansible job connecting to hundreds of hosts to roll out an update) and I think the events associated with that long-running job being associated with different bot instances would be confusing. 

> And I imagine a decent share of the cloud provider / k8s joins that could be stateful end up being ephemeral anyway.

Right now, there's not much of a motivation to have state configured because there's no real advantage to it, in fact, at one point there was a disadvantage as an expired cert would cause the bot to be unable to start. I'd imagine that if we were to introduce BotInstance's then there's a perceived benefit at least in being able to attribute actions to a particular instance of tbot throughout its lifetime in a fairly persistent way.

I suppose the other overriding thought in my head is I'm desperate to stop the token vs delegated join method behaviour diverting further. It's already confusing enough and I can see user frustration when the less secure join method is actually delivering more features to them (in terms of having a long-lived persistence identity they can attribute actions to).

- Add support for calling the join RPCs with a client certificate.

> This technique could reuse a recently proposed LoginID attribute. This would 

Author: I'm curious about whether we'll move forward with the proposed idea of LoginIDs from the hackathon - we could even straight up re-use this for join/renews of Machine ID.

Author: It looks like there's not really been much movement on this since January. We can probably safely assume we do not want to depend on that.

Reviewer: Since it looks like we're dropping this, I'll remove the reference.

  // method, this counter is checked during renewal and the Bot is locked out if
  // the counter in the certificate does not match the counter of the last
  // authentication.
  int32 generation = 5;
}

// BotInstanceStatus holds the status of a BotInstance.
message BotInstanceStatus {
  // The public key of the Bot instance.
  // When authenticating a Bot instance, the full public key must be compared
  // rather than just the fingerprint to mitigate pre-image attacks.
  bytes public_key = 1;
  // The fingerprint of the public key of the Bot instance.
  string fingerprint = 2;
  // The name of the Bot that this instance is associated with.
  string bot_name = 3;
  // Last X records kept, with the second oldest being removed once the limit
  // is reached. This avoids the indefinite growth of the resource but also
  // ensures the initial record is retained.
  repeated BotInstanceStatusAuthentication authentications = 4;
  // Last X records kept, with the second oldest being removed once the limit
  // is reached. This avoids the indefinite growth of the resource but also
  // ensures the initial record is retained.

>   repeated BotInstanceStatusHeartbeat heartbeats = 5;

Author: Sure - feel free to make this change. I don't really have a strong opinion about how this is stored.

  comes from the same instance in time.
- Allows self-reported information to be used as part of renewal decision.
  This is not a strong defence as it is self-reported and cannot be trusted.
- Avoids a state where the BotInstance is incomplete immediately after joining
  and before it has called SubmitHeartbeat.

Cons:

- Adds Bot specific behaviour to RPCs that are also used for Node joining.
- Heartbeats are limited to the interval of renewal.

#### API

Additional RPCs will be added to the BotInstance service to allow these to
be listed and deleted:

```protobuf
syntax = "proto3";

package teleport.machineid.v1;

service BotInstanceService {
  // GetBotInstance returns the specified BotInstance resource.
  rpc GetBotInstance(GetBotInstanceRequest) returns (BotInstance);
  // ListBotInstances returns a page of BotInstance resources.
  rpc ListBotInstances(ListBotInstancesRequest) returns (ListBotInstancesResponse);
  // DeleteBotInstance hard deletes the specified BotInstance resource.
  rpc DeleteBotInstance(DeleteBotInstanceRequest) returns (google.protobuf.Empty);
}

// Request for GetBotInstance.
message GetBotInstanceRequest {
  // The name of the BotInstance to retrieve.

>   string name = 1;

Author: I wonder if it makes more sense to seperate the bot name and fingerprint fields here rather than provide them concatenated.

Author: Yeah -  I think I'm growing to like the idea of explicit fields - but I do see one potential issue. If I already have a BotInstance resource, it's very easy to call GetBotInstance using the name of the resource in this field if I want to  check for any updates. If we switch to separate fields, there's potentially a little more complexity in this as you'd have to do something like

```
GetBotInstance({
  Bot: instance.spec.bot,
  Instance: instance.name
})
```

This feels fairly minor so I haven't got a super strong opinion here. Feel free to change.

Reviewer: I've split the fields, though on some level I wonder if we could just refer directly to the bot instance identifier. The bot name namespaces the UUID, but is that really necessary?

Author: Hm - but with only the UUID, we won't be able to get the BotInstance from the store, as we're going to include the bot name in the storage key to enable listing of BotInstances for a given bot.

Reviewer: Ah, good call, didn't think about that.

  // The server may impose a different page size at its discretion.
  int32 page_size = 2;
  // The page_token value returned from a previous ListBotInstances request, if
  // any.
  string page_token = 3;
}

// Response for ListBotInstances.
message ListBotInstancesResponse {
  // BotInstance that matched the search.
  repeated BotInstance bot_instances = 1;
  // Token to retrieve the next page of results, or empty if there are no
  // more results exist.
  string next_page_token = 2;
}

// Request for DeleteBotInstance.
message DeleteBotInstanceRequest {
  // The name of the BotInstance to delete.
  string name = 1;
}
```

### Changes to the `token` Join Method

As we now have a way to track the generation for a specific Bot instance, we
can allow multiple Bot instances to be associated with a single Bot. This
also means that the token no longer needs to be consumed on a join.

Eventually, we may wish to add a way to specify a number of joins which can
occur with a token. This provides a way to easily control the lifetime of a 
token when deploying to a fleet of a pre-known size.

The renewal logic will need to be adjusted to read and update the generation

> counter from the BotInstance rather than the Bot user.

Author: > I'm a bit conflicted about launching multi-use tokens without also adding options to limit joins.

I agree some sort of limiting factor on a number of joins is likely going to be necessary out of the box.

> We should include a join count limit from the start, and arguably default it to 1 so users are making an explicit decision to allow multiple bots to join. This default would also avoid a potentially surprising change in behavior for existing users.

I like this idea - especially using a default of 1 to avoid a big change in behaviour.

---

One use-case I think we can't neglect is CI platforms we do not support and can not support due to their limitations. In this case, users may well still want a token that is "quite" long-lived and supports a large number of joins.

We'd still want them to rotate though... one thing I was thinking about to make this easier would be to produce a script or command which can be executed on a cron job that generates a new token for a named bot and then calls a user-supplied script which should connect to their CI platform/secret store and updates the token. This may be a nice answer for those who are backed into a corner by a rudimentary CI platform.

This fits well into the "bot refresh" idea, some sort of easyish CLI command where you specify an existing bot and can just generate a new token.

> counter from the BotInstance rather than the Bot user.

Reviewer: > One use-case I think we can't neglect is CI platforms we do not support and can not support due to their limitations. In this case, users may well still want a token that is "quite" long-lived and supports a large number of joins.

I agree. I almost wonder if there's a middle ground here where we might be able to force users to make a security decision:
* The `token` join method sets a sane cap on the TTL of a join token to proactively encourage some variety of automation, say 7 days or something. I don't think we need to cap the number of uses as long as there's a time bound.
  * It might be useful to write up a cron job example, maybe a guide on using a GHA or k8s cronjob to rotate a Jenkins joining secret or something.
* We then (eventually) introduce a new `token-insecure` that otherwise behaves identically but has no restrictions at all on TTL or number of joins (basically this [other thread](https://github.com/gravitational/teleport/pull/36510/files#r1602429025) down the page). This method might also opt out of the generation counter.

I think we could argue enforcing a maximum token TTL is an acceptable breaking change given how narrowly useful long TTL'd tokens are today, and then introduce `token-insecure` at some point in the future? I can't imagine any users today are making practical use of extremely long TTL bot tokens that would have their use cases broken.

> counter from the BotInstance rather than the Bot user.

Author: I'm a little cautious about setting a cap on the TTL, if we were totally greenfield I'd agree but:
- Today, there's no cap. I agree with you and doubt anyone is relying on this behaviour - so perhaps we can discount this.
- We'd need to have logic that ignores the cap if it's for Agent joining
- We'd need to have logic that ignores the cap if the join method is not `token`.

I'd perhaps be less cautious about the TTL cap if we were approaching this at a later date when we add `token-insecure`, imho, the whole token type could do with a V3 and that might be a good opportunity to do so. As an aside, I'd love to make it so token name is not the actual secret value, e.g a `token` join method token has a UUID-esque name or a user provided name, and then a field which specifies the actual secret value. This would make audit logging etc much easier because we wouldn't need to "hide" the name.

I'm not overly decided on this and I think I need to ponder on it a little longer to figure out my full thoughts. It's almost out of scope for this RFD - so perhaps we don't need to make a decision on this immediately and we can proceed with the BotInstance work since I don't think this has any impact.

> counter from the BotInstance rather than the Bot user.

Reviewer: I think my hopefully more in scope recommendation for this RFD is: let's enforce a soft client-side limit in `tctl` and/or warn that it's a bad idea and we may restrict token length in the future. I figure that if a user wants to create an extremely long-lived token, they can write up the YAML themselves and `tctl create -f` it.

I think warnings or soft errors today help keep the door open for us to lock this down more in the future while making sure use cases that need very long lived tokens are at least possible.

single counter within the Bot User labels, we could store a JSON encoded map
of counters.

One challenge would be contention over the user resource if a large number of
Bot instances are trying to renew their certificates at the same time. Our
Backend has limited support for transactional consistency and this increases the
risk of two Bot instances renewing simultaneously and producing an inconsistent
state that locks one of them out.

### Introduce renewal-less and generation-less `token` join method

One option is to introduce a new join method that does not produce renewable
certificates. There would be no need for the fragile generation counter
and the join token would be continually re-used to join as is done for the
delegated join methods. This also circumvents the need for a one-to-one binding
between a Bot instance and a Bot.

This token would be incredibly sensitive and if stolen, there would be no
automated mechanisms to detect this as exists today with the generation counter.

It likely makes more sense to improve the existing `token` join method rather
than introduce a variant which behaves differently and is less secure. It would

> increase the complexity of the codebase and the user experience.

Author: Definitely out of scope for this quarter - but well worth investigating I should think.

## Out of Scope

These tasks are out of scope of this RFD but could be considered natural 
follow-on tasks.

### Multi-phase Commit of Generation Counter

Currently, the generation counter is fragile as it is incremented server side
without confirmation that `tbot` has been able to use and persist the new
credentials. If `tbot` does not receive confirmation of the renewal or is
unable to persist the new credentials, it will be locked out on it's next
attempt to renew.

We could introduce a multi-phase commit of the generation counter. This would
provide more robustness to the renewal process.

### Locking of Individual Bot Instances

Currently, it's only possible to lock out an entire Bot user. This means that
when managing a large fleet, it would not be able to lock out a specific host
that had been compromised. This is likely to be a major friction point for those
deploying a large number of Bot instances.

> It also increases the significance of the fragility of the generation counter.

Author: One thing I thought about here is an option to "Lock all bot instances that have authenticated with X token". Seems fairly feasible if we store the hash of the token used on the first join or similar.

I think this is deffo something for next quarter though.

---
authors: Noah Stride (noah.stride@goteleport.com)
state: draft
---

# RFD 00162 - Improving the on-prem fleet Bot management experience

## Required approvers

- Engineering: @zmb3
- Product: @klizhentas || @xinding33
- Security: @reedloden || @jentfoo

## What

Terminology:

- Bot: An identity within Teleport intended for use by machines as opposed to
  humans. Many individual machines may act as this shared identity.

> - `tbot`: The Teleport binary that acts as aBot and generates credentials for

Reviewer 3: ```suggestion
- `tbot`: The Teleport binary that acts as a Bot and generates credentials for
```

  consumption by client applications.
- Bot instance: A single instance of `tbot` running on a host.

This RFD proposes improvements to the management of fleets of Machine ID Bots.
These improvements are mostly targetted at on-prem deployments, where the
delegated join methods are not available.

The improvements will focus on three points:

- To allow multiple Bot instances to be associated with a single Bot when using
  the `token` join method.
- To allow multiple Bot instances to be joined using a single join token when
  using the `token` join method.
- Providing a way to track and monitor Bot instances.

## Why

> Whilst deploying a large fleet of Bots is fairly trivial when using the

Reviewer 3: I might mention TPM joining here as well. If TPMs are available, that flow is probably preferred.

delegated join methods, the experience when managing a fleet of bot hosts
on-prem is more challenging.

The following burdens currently exist:

> - When using the `token` join method, a Bot must be created for each host. This

Reviewer 3: ```suggestion
- When using the `token` join method, a Bot must be created for each Bot instance. This
```

To reuse the terminology you recently introduces.

  means that the privileges of many distinct Bots need to be synchronised where
  those hosts are performing the same function.
- When using the `token` join method, a token can only be used once. This means
  creating hundreds of join tokens and managing securely distributing these to
  hosts.

> - When managing a large fleet of `tbot` deployments, there is no way to

Reviewer 3: ```suggestion
- When managing a large fleet of Bot instances, there is no way to
```

during the join process should accomplish both of these goals.

##### Verifying Bot Identities

We currently see two methods for cryptographically verifying bot identities at
renewal time:

1. We could expose the existing functionality of the HTTPS-only
   `RegisterUsingToken` over gRPC. The existing gRPC `JoinService` can be
   accessed with and without authentication, so we could inspect the client
   connection to find the existing bot identity, if any, and it would be
   implicitly verified.
2. We could adapt the existing HTTPS implementation of `RegisterUsingToken` to
   additionally accept an encoded existing certificate, and return certificates
   encrypted with the certificate's public key. We can verify the certificate
   was originally signed with our CA, and the client will only be able to
   decrypt the returned identity if they actually have the private key for the
   previous identity.

Our preference is option (1): bots always join over gRPC, and provide a client
certificate when re-joining. Bots without a client cert to present are
registered as new instances, while bots with a valid client cert preserve their

> identity.

Reviewer 3: +1, nice and simple

will be followed.

With a persistent identifier for a Bot instance established, we can now track
information about a specific Bot instance server-side. In addition to providing
a way to store a generation counter per instance, this yields other benefits:

- Allows Bot instances to be viewed within the UI and CLI.
- Allowing Bot instances to submit basic self-reported information about itself
  and its host, e.g:
  - `tbot` version
  - Hostname, OS and OS version
  - The configuration of `tbot`
  - Health status
- Record metadata from delegated joins to enrich information about the Bot.
  E.g show the linked repository / CI run number
- Billing based on Bot instances rather than Bots.

Some of this information is known and verified by the server - for example, the
certificate generation or the join metadata. Some of this information is
self-reported and should not be trusted. The information from these two sources
should be segregated to avoid confusion.

BotInstance will be a new resource type introduced to track this information.

```protobuf
syntax = "proto3";

import "teleport/header/v1/metadata.proto";

// A BotInstance

> message BotInstance {

Reviewer 3: Does this resource have any impact on our Terraform provider?

Reviewer: I don't think so, bot instances will be created automatically when a bot joins. They're more for record keeping and visibility than something end users need to configure.

I could see exposing it in the future if we add some command and control functionality to `BotInstanceSpec`, though.

  // The fingerprint of the public key of the Bot instance.
  string fingerprint = 7;
}

// BotInstanceStatus holds the status of a BotInstance.
message BotInstanceStatus {
  // The unique identifier for this bot.
  string id = 1;
  // The name of the Bot that this instance is associated with.
  string bot_name = 2;
  // The initial authentication status for this bot instance.
  BotInstanceStatusAuthentication initial_authentication = 3;
  // The N most recent authentication status records for this bot instance.
  repeated BotInstanceStatusAuthentication latest_authentications = 4;
  // The initial heartbeat status for this bot instance.
  BotInstanceStatusHeartbeat initial_heartbeat = 5;
  // The N most recent heartbeats for this bot instance.
  repeated BotInstanceStatusHeartbeat latest_heartbeats = 6;
}
```

The name used for a BotInstance will be a concatenation of the Bot name and its
unique identifier (UUID).

When storing the BotInstance in the backend, the key will be:
`bot_instances/{bot_name}/{uuid}`. This will allow for efficient listing
of BotInstances for a given Bot.

Like agent heartbeats, the BotInstance will expire after a period of inactivity.

> This avoids the accumulation of ephemeral BotInstances.

Reviewer 3: Do we anticipate large numbers of bot instances at scale? Do we need to do something like variable rate heartbeats here, or is a fixed TTL reasonable?

Reviewer: I'd guess a customer will eventually want to create a huge number of bot instances. What do you mean by variable rate, would jitter qualify?

behaviour.

#### Recording Heartbeat Data

A new RPC will be added for submitting heartbeat data:

```protobuf
syntax = "proto3";

package teleport.machineid.v1;

service BotInstanceService {
  // SubmitHeartbeat submits a heartbeat for a BotInstance.
  rpc SubmitHeartbeat(SubmitHeartbeatRequest) returns (SubmitHeartbeatResponse);
}

// The request for SubmitHeartbeat.
message SubmitHeartbeatRequest {
  // The heartbeat data to submit.
  BotInstanceStatusHeartbeat heartbeat = 1;
}

// The response for SubmitHeartbeat.
message SubmitHeartbeatResponse {
  // Empty
}
```

The endpoint will have a special auth/authz check. RBAC will not be used and
instead the endpoint will ensure that the presented client certificate is for
the Bot linked to the instance.

This endpoint will be called by `tbot` immediately after it has initially
authenticated. After a heartbeat has successfully completed, another should be
scheduled for a half hour after. A small amount of jitter should be added to
the heartbeat period to avoid a thundering herd of heartbeats.

If the heartbeat fails, then `tbot` should retry on an exponential backoff.

> ##### Alternative: Submit Heartbeat Data on Join/Renew

Reviewer 3: Before merging this, it would be helpful if each `Alternative` that we present but ultimately do not select lists why it was not chosen.

Reviewer: It's been long enough now that I don't remember the order of events, but we now have a summary of why alternatives were not selected for (at least) public key fingerprints and the heartbeat RPC.

message DeleteBotInstanceRequest {
  // The name of the BotInstance to delete.
  string bot_name = 1;
  // The unique identifier of the bot instance to delete.
  string id = 2;
}
```

### Changes to the `token` Join Method

As we now have a way to track the generation for a specific Bot instance, we
can allow multiple Bot instances to be associated with a single Bot. This
also means that the token no longer needs to be consumed on a join.

However, this does introduce a change in our security guarantees, and without
additional tooling support and sensible defaults, the change may incentivize end
users to create long-lived join tokens instead of using a more appropriate join
method, or automating issuance of short lived tokens.

To this end, we should introduce a per-bot-instance join count limit, and
configure that to be 1 join by default. This matches today's behavior, and
will help ensure users do not accidentally create a token that provides more
access than expected: "infinite use" tokens with massive join limits and/or very
long TTLs will need to be explicitly specified.

We may additionally want to put hurdles in the way of things like extremely long

> token TTLs. There are legitimate low-security use cases for these, but we could

Reviewer 3: I'm curious what Sasha thinks about this.

Extremely long join token TTLs basically become a shared secret like an API key, and we built Machine ID to avoid long-lived shared secrets..

Reviewer: I agree, and would ideally like to only support this by way of an explicitly insecure `token-insecure` join method at some point in the future (see previous thread: https://github.com/gravitational/teleport/pull/36510#discussion_r1604165929). As we discussed in that thread, adding a TTL limit would _technically_ be a breaking change, and there are use cases we may still want to support.

For context on why, today's status quo is problematic because it seems like we properly support token joining, but it's kind of an illusion. It works great in a Getting Started guide on your dev box, and then breaks as soon as you try it in production because one of your nodes went offline for more than an hour. Production use really requires full automation (e.g. ansible or terraform detecting dead bots and reissuing tokens), and I'm not aware of any customers who've actually put in the effort to do it.

The changes proposed here (multiple uses per token, multiple tokens per bot) simplify the automation requirements a lot, but it's still a heavy lift, and I'm not sure it'll ever be worthwhile for a lot of use cases. There may be value in giving customers in this niche a way to still make use of Teleport if they (explicitly) accept the security implications of a long lived token; they still benefit from our RBAC, audit logging, etc.

Reviewer 4: Michael is working on the display and creation of join tokens in the web UI, and we actually _just_ had a conversation with Sasha about token creation during the Design and Product Office Hours yesterday. In the meeting, we agreed to show a danger state for any `token` join method with an expiration >24 hours. We also agreed to make it impossible to create long-lived tokens through the web UI. 

(What we discussed doesn't affect the underlying mechanics of the `token` join method, but it does put some guard rails for their creation in the web UI and affects how they will be _displayed,_ regardless of how they were created.)

introduce a soft limit in `tctl` preventing automatic token creation with TTL
longer than 7 days, forcing users to manually create the token if they really
need it to last longer.

Additionally, renewal logic will need to be adjusted to read and update the
generation counter from the BotInstance rather than the Bot user. We should also
take care to ensure this counter behaves sensibly even when many bots are
attempting to join the cluster concurrently. The generation counter today
already has issues with concurrent joins, and it's even more important to get
this right when we can expect contention over a single bot instance resource.

### CLI Changes

#### `tbot`

`tbot reset` will be added to allow a `tbot` instance to be reset. This will
simply clear out any artifacts within the `tbot` storage directory.

In addition, if the bot detects a change in join token or join method, it should

> automatically rotate it's keypair. This will ensure it presents as a fresh

Reviewer 3: ```suggestion
automatically rotate its keypair. This will ensure it presents as a fresh
```

But you want to do this in a way that it can still use its existing client certs to re-join, right?

Reviewer: Yes, and we'll be supporting rejoin-then-rotate regardless. I think since we record the token and join method as part of the last <n> authentication records, it makes sense for that to be our behavior if bot config changes.

BotInstance to the AuthServer.

A log message should be output that identifies the linked BotInstance at
startup and on each heartbeat. This will allow users to easily correlate the
`tbot` installation with a BotInstance.

#### `tctl`

Commands to list all BotInstances and the BotInstances for a specific Bot should
be added to the `tctl bots` family:

> `tctl bots instances list`

Reviewer 3: Do we need `instances`? What about `tctl bots ls`?

Reviewer: I think the two are distinct, `tctl bots ls` will continue to work as it does today, showing (effectively) a name + allowed roles. You'd then list all the unique joins of a given bot with `tctl bots instances list --bot <name>`.

There's some overlap, I think we could amend `tctl bots ls` to say, show a unique count of instances per bot, for example.

Author: Yeah - that's correct, we'll still want way to list the actual bot identities.

`tctl bots instances list --bot <bot name>`

The `tctl tokens add` command should be extended to allow a new token to be
associated with an existing Bot now that multiple Bot instances can be run
against a single Bot:

`tctl tokens add --type=bot --bot <bot name>`

Additionally, `tctl rm`/`tctl get` should be able to operate on BotInstance.

There is no requirement for it to be possible to create or update a BotInstance
with `tctl`.

### Analytics

A PostHog event should be emitted for each BotInstance heartbeat. This will

> allow us to track active bots in a similar way to how we track active agents.

Reviewer 3: Should we included it in the aggregated `tp.resource.heartbeat` event rather than creating a new event for each heartbeat?

  string version = 2;
  // indicates whether or not tbot is running in one-shot mode
  bool one_shot = 3;
  // indicates the configured join method of `tbot`.
  string join_method = 4;
}
```

Existing analytics for join, renewal and certificate generation should be
extended to include the BotInstance ID anonymized. This will allow them to be
linked together.

### Migration/Compatability

The "create if not exists" behaviour of the BotInstance resource will mean that
existing Bot instances will have a BotInstance resource created on their first
renewal after this feature is released. Their existing generation counter will
be trusted on this first renewal. This allows for a seamless migration to the
new system.

Older `tbot` instances will not submit heartbeats. This means that their
BotInstance will only contain authentication data. Any CLI or GUI that shows
BotInstances should show a gracefully degraded state in this case that explains
that the `tbot` needs to be upgraded.

## Security Considerations

### Audit Events

An audit event should be added for the deletion of a BotInstance. The
name of the BotInstance should be added to the existing join, renewal and

> certificate generation audit events.

Reviewer 3: We should also put the bot instance's name/UUID in any existing audit events (like certificates issued) so that we have traceability into the actions of a particular bot.

In the future, we can then extend the access monitoring feature or even offer a bot details page in the UI that lists its renewals, join attempts, certs issued, etc.

This token would be incredibly sensitive and if stolen, there would be no
automated mechanisms to detect this as exists today with the generation counter.

It likely makes more sense to improve the existing `token` join method rather
than introduce a variant which behaves differently and is less secure. It would
increase the complexity of the codebase and the user experience.

## Out of Scope

These tasks are out of scope of this RFD but could be considered natural
follow-on tasks.

### Multi-phase Commit of Generation Counter

Currently, the generation counter is fragile as it is incremented server side
without confirmation that `tbot` has been able to use and persist the new
credentials. If `tbot` does not receive confirmation of the renewal or is
unable to persist the new credentials, it will be locked out on it's next
attempt to renew.

We could introduce a multi-phase commit of the generation counter. This would
provide more robustness to the renewal process.

### Locking of Individual Bot Instances

Currently, it's only possible to lock out an entire Bot user. This means that
when managing a large fleet, it would not be able to lock out a specific host

> that had been compromised. This is likely to be a major friction point for those

Reviewer 3: We can leave this out of scope but we should explore it in the future.

We can already lock any heartbeating agent based on its Host ID today, so it feels like this should extend to bots without too much trouble.

Reviewer: Yeah, we should add _something_ here, but it's trivial to circumvent when the join token can just be reused, @strideynet had some ideas here: https://github.com/gravitational/teleport/pull/36510/files/61e17365cac7aaec059de86d470a57a6b612eebd#r1602421886

2. We could adapt the existing HTTPS implementation of `RegisterUsingToken` to
   additionally accept an encoded existing certificate, and return certificates
   encrypted with the certificate's public key. We can verify the certificate
   was originally signed with our CA, and the client will only be able to
   decrypt the returned identity if they actually have the private key for the
   previous identity.

Our preference is option (1): bots always join over gRPC, and provide a client
certificate when re-joining. Bots without a client cert to present are
registered as new instances, while bots with a valid client cert preserve their
identity.

Note that bots joined with the `token` method are not affected by this, as
renewals take place over a fully mTLS-authenticated tunnel to the auth service.

Additionally, a downside to certificate verification is that the certificate
validity period becomes a factor. If bots are only run intermittently, like
from a CI workflow, their certificates could expire and prevent them from
being identified as the same instance. This is likely to only impact a small
number of cases, however, as most CI provider joins are stateless and have no

> certificates to present anyway. Bots that present expired certificates will

Author: I imagine the TLS handshake will reject them for us - we can just make sure the client is capable of handling this and will try without a certificate.

  // Server.
  google.protobuf.Timestamp authenticated_at = 1;
  // The join method used for this join or renewal.
  string join_method = 2;
  // The join token used for this join or renewal. This is only populated for
  // delegated join methods as the value for `token` join methods is sensitive.
  string join_token = 3;
  // The metadata sourced from the join method.
  google.protobuf.Struct metadata = 4;
  // On each renewal, this generation is incremented. For delegated join
  // methods, this counter is not checked during renewal. For the `token` join
  // method, this counter is checked during renewal and the Bot is locked out if
  // the counter in the certificate does not match the counter of the last
  // authentication.
  int32 generation = 5;
}

// BotInstanceStatus holds the status of a BotInstance.
message BotInstanceStatus {
  // The unique identifier for this bot.
  string id = 1;
  // The public key of the Bot instance.
  // When authenticating a Bot instance, the full public key must be compared
  // rather than just the fingerprint to mitigate pre-image attacks.
  bytes public_key = 2;
  // The fingerprint of the public key of the Bot instance.

>   string fingerprint = 3;

Reviewer: I've so far kept the public key and fingerprint fields even though we're adding a UUID. Do we want to keep these? I don't think they're strictly required, but may still be useful.

Author: I think it's useful info but if it's placed at the top level, we don't have a context point for when the value was this. We could move this into `BotInstanceStatusAuthentication` and that way we'll properly handle the key changing.

Reviewer: Good call, will move it into that message.

  string bot_name = 4;
  // The initial authentication status for this bot instance.
  BotInstanceStatusAuthentication initial_authentication = 5;
  // The N most recent authentication status records for this bot instance.
  repeated BotInstanceStatusAuthentication authentications = 6;
  // The initial heartbeat status for this bot instance.
  BotInstanceStatusHeartbeat initial_heartbeat = 7;
  // The N most recent heartbeats for this bot instance.
  repeated BotInstanceStatusHeartbeat latest_heartbeats = 8;
}
```

The name used for a BotInstance will be a concatenation of the Bot name and its
unique identifier (UUID).

When storing the BotInstance in the backend, the key will be:
`bot_instances/{bot_name}/{uuid}`. This will allow for efficient listing
of BotInstances for a given Bot.

Like agent heartbeats, the BotInstance will expire after a period of inactivity.
This avoids the accumulation of ephemeral BotInstances.

#### Recording Authentication Data

Upon each join and renewal, the BotInstance record will be updated with an
additional entry in the `status.authentications` field. If there is X entries,
then the second-oldest entry will be removed. This prevents growth without

> bounds but also ensures that the original record is retained.

Author: Now that we're storing the initial as it's own field, we will want to delete the oldest rather than second oldest.

> In addition, the TTL of the BotInstance resource will be extended.

Author: ```suggestion
In addition, the TTL of the BotInstance resource will be extended to cover the validity period of the issued certificate, plus a short additional time to allow for some imprecision.
```

If a BotInstance does not exist, then one will be created. In the case that
this occurs for a bot using the `token` join method and this is a renewal,
a warning will be emitted and the initial generation of the BotInstance will
be sourced from the certificates current generation counter. This behaviour
will support the migration of existing `tbot` instances to the new BotInstance
behaviour.

#### Recording Heartbeat Data

A new RPC will be added for submitting heartbeat data:

```protobuf
syntax = "proto3";

package teleport.machineid.v1;

service BotInstanceService {
  // SubmitHeartbeat submits a heartbeat for a BotInstance.
  rpc SubmitHeartbeat(SubmitHeartbeatRequest) returns (SubmitHeartbeatResponse);
}

// The request for SubmitHeartbeat.
message SubmitHeartbeatRequest {
  // The heartbeat data to submit.
  BotInstanceStatusHeartbeat heartbeat = 1;
}

// The response for SubmitHeartbeat.
message SubmitHeartbeatResponse {
  // Empty
}
```

The endpoint will have a special auth/authz check. RBAC will not be used and
instead the endpoint will check:

- The presented client certificate is for the Bot linked to the instance.

> - The presented client certificate's public key matches the public key recorded

Author: We can drop this requirement now that we're using UUIDs.

- Allows self-reported information to be used as part of renewal decision.
  This is not a strong defence as it is self-reported and cannot be trusted.
- Avoids a state where the BotInstance is incomplete immediately after joining
  and before it has called SubmitHeartbeat.

Cons:

- Adds Bot specific behaviour to RPCs that are also used for Node joining.
- Heartbeats are limited to the interval of renewal.

#### API

Additional RPCs will be added to the BotInstance service to allow these to
be listed and deleted:

```protobuf
syntax = "proto3";

package teleport.machineid.v1;

service BotInstanceService {
  // GetBotInstance returns the specified BotInstance resource.
  rpc GetBotInstance(GetBotInstanceRequest) returns (BotInstance);
  // ListBotInstances returns a page of BotInstance resources.
  rpc ListBotInstances(ListBotInstancesRequest) returns (ListBotInstancesResponse);
  // DeleteBotInstance hard deletes the specified BotInstance resource.
  rpc DeleteBotInstance(DeleteBotInstanceRequest) returns (google.protobuf.Empty);
}

// Request for GetBotInstance.
message GetBotInstanceRequest {
  // The name of the bot associated with the instance.

>   string name = 1;

Author: ```suggestion
  string bot_name = 1;
```

---
authors: Noah Stride (noah.stride@goteleport.com)

> state: draft

Author: ```suggestion
state: Implemented (16.2.0)
```
